ocaml-lsp-server/src/custom_requests/req_hover_extended.ml

```diff
@@ -0,0 +1,116 @@
+    ; cursor_position : Position.t
+    ; verbosity : int
+    }
+
+  let params_schema =
+    `Assoc
+      [ ("textDocument", `String "<TextDocumentIdentifier>")
+      ; ("position", `String "<Position>")
+      ; ("verbosity", `String "<integer>")
+      ]
+
+  let of_jsonrpc_params params : t option =
+    match params with
+    | `Assoc
+        [ ("textDocument", text_document)
+        ; ("position", position)
+        ; ("verbosity", verbosity)
+        ] ->
+      let text_document = TextDocumentIdentifier.t_of_yojson text_document in
+      let cursor_position = Position.t_of_yojson position in
+      let verbosity = Yojson.Safe.Util.to_int verbosity in
+      Some { text_document; cursor_position; verbosity }
+    | _ -> None
+
+  let of_jsonrpc_params_exn params : t =
+    let params_spec = { Custom_request.params_schema; of_jsonrpc_params } in
+    Custom_request.of_jsonrpc_params_exn params_spec params
+end
+
+let client_capabilities (state : State.t) =
+  (State.initialize_params state).capabilities
+
+(* TODO: duplicated with Hover_req.format_contents *)
+let format_contents ~syntax ~markdown ~typ ~doc =
+  (* TODO for vscode, we should just use the language id. But that will not work
+     for all editors *)
+  `MarkupContent
+    (if markdown then
+     let value =
+       let markdown_name = Document.Syntax.markdown_name syntax in
+       match doc with
+       | None -> sprintf "```%s\n%s\n```" markdown_name typ
+       | Some s ->
+         let doc =
+           match Doc_to_md.translate s with
+           | Raw d -> sprintf "(** %s *)" d
+           | Markdown d -> d
+         in
+         sprintf "```%s\n%s\n```\n---\n%s" markdown_name typ doc
+     in
+     { MarkupContent.value; kind = MarkupKind.Markdown }
+    else
+      let value =
+        match doc with
+        | None -> sprintf "%s" typ
+        | Some d -> sprintf "%s\n%s" typ d
+      in
+      { MarkupContent.value; kind = MarkupKind.PlainText })
+
+let hover (* server *) (state : State.t) text_document position verbosity =
+  let doc =
+    Document_store.get state.store text_document.TextDocumentIdentifier.uri
+  in
+  let pos = Position.logical position in
+  (* TODO we shouldn't be acquiring the merlin thread twice per request *)
+  let* type_enclosing = Document.type_enclosing ~verbosity doc pos in
+  match type_enclosing with
+  | None -> Fiber.return None
+  | Some { loc; typ; doc = documentation } ->
+    let syntax = Document.syntax doc in
+    let+ typ =
+      (* We ask Ocamlformat to format this type *)
+      let* result = Ocamlformat_rpc.format_type state.ocamlformat_rpc ~typ in
+      match result with
+      | Ok v ->
+        (* OCamlformat adds an unnecessay newline at the end of the type *)
+        Fiber.return (String.trim v)
+      | Error `No_process -> Fiber.return typ
+      | Error (`Msg _message) ->
+        (* We log OCamlformat errors and display the unformated type *)
+        (* TODO: how to obtain the server *)
```

Comment:
I could add the server as an arg to all the custom commands

https://github.com/ocaml/ocaml-lsp/blob/259bf6e0a016d209c719ab9adecf1f88c072e71a/ocaml-lsp-server/src/ocaml_lsp_server.ml#L720-L725